lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -305,6 +305,266 @@ class ConvertValsemVariantAtenIndexPutImplOp
+    w = indicesShape[tensorOperandRank - 1];
+    hTimesW = indicesShape[tensorOperandRank - 2] * w;
+    cTimesHTimesW = indicesShape[tensorOperandRank - 3] * hTimesW;
+
+    SmallVector<AffineExpr, 4> dimExprs;
+    for (int64_t i = 0; i < tensorOperandRank; i++)
+      dimExprs.push_back(rewriter.getAffineDimExpr(i));
+    SmallVector<AffineExpr, 2> updatedIndicesExpr;
+    if (tensorOperandRank == 4) {
+      updatedIndicesExpr.push_back(dimExprs[0] * cTimesHTimesW +
+                                   dimExprs[1] * hTimesW + dimExprs[2] * w +
+                                   dimExprs[3]);
+    } else {
+      updatedIndicesExpr.push_back(dimExprs[0] * hTimesW + dimExprs[1] * w +
+                                   dimExprs[2]);
+    }
+    updatedIndicesExpr.push_back(rewriter.getAffineDimExpr(tensorOperandRank));
+
+    SmallVector<AffineMap> indexingMaps =
+        AffineMap::inferFromExprList({dimExprs, updatedIndicesExpr});
+    SmallVector<StringRef> iteratorTypes(tensorOperandRank + 1, "parallel");
+
+    Value initTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, llvm::makeArrayRef({numelIndices, tensorOperandRank}),
+        indicesElemType);
+
+    Value cstWin = rewriter.create<arith::ConstantIndexOp>(
+        loc, /*Win=*/inputShape[tensorOperandRank - 1]);
+    Value cstIndexZero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value cstIndexOne = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value cstIndexTwo = rewriter.create<arith::ConstantIndexOp>(loc, 2);
+    Value cstIndexThree = rewriter.create<arith::ConstantIndexOp>(loc, 3);
+
+    Value updatedIndices =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, initTensor.getType(), indices, initTensor, indexingMaps,
+                iteratorTypes,
+                [tensorOperandRank, cstIndexZero, cstIndexOne, cstIndexTwo,
+                 cstIndexThree, cstWin,
+                 indicesElemType](OpBuilder &b, Location loc, ValueRange args) {
+                  Value index = castIntToIndex(b, loc, args[0]), outIndex;
+                  Value lastDim =
+                      b.create<linalg::IndexOp>(loc, tensorOperandRank);
+
+                  Value predZero = b.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::eq, lastDim, cstIndexZero);
+                  Value predOne = b.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::eq, lastDim, cstIndexOne);
+                  Value predTwo = b.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::eq, lastDim, cstIndexTwo);
+                  Value predThree = b.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::eq, lastDim, cstIndexThree);
+
+                  Value outIndexDimZero = b.create<arith::SelectOp>(
+                      loc, predZero, b.create<linalg::IndexOp>(loc, 0),
+                      cstIndexZero);
+
+                  if (tensorOperandRank == 4) {
```

Comment:
the 2 if branches are very similar, can you try to refactor them to deduplicate the code? Something like:
```
sum = 0
for i in Rank-1..0:
  pred = pred(cst(i))
  if (i == Rank-1)
      result = index%Win
  if (i == Rank-2)
      result = index/Win
  if (i == Rank-3 || i == Rank-4)
      result = linalg.index(i)
  addAmount = select(pred, amount, 0)
  sum += addAmount
```